framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1050,34 +1054,78 @@ protected void checkPurity(MethodTree tree) {
       if (TreeUtils.isConstructor(tree) || TreeUtils.isVoidReturn(tree)) {
         additionalKinds.remove(Pure.Kind.DETERMINISTIC);
       }
-      if (!additionalKinds.isEmpty()) {
-        if (infer) {
-          WholeProgramInference wpi = atypeFactory.getWholeProgramInference();
-          ExecutableElement methodElt = TreeUtils.elementFromDeclaration(tree);
-          if (additionalKinds.size() == 2) {
-            wpi.addMethodDeclarationAnnotation(methodElt, PURE);
-          } else if (additionalKinds.contains(Pure.Kind.SIDE_EFFECT_FREE)) {
-            wpi.addMethodDeclarationAnnotation(methodElt, SIDE_EFFECT_FREE);
-          } else if (additionalKinds.contains(Pure.Kind.DETERMINISTIC)) {
-            wpi.addMethodDeclarationAnnotation(methodElt, DETERMINISTIC);
-          } else {
-            throw new BugInCF("Unexpected purity kind in " + additionalKinds);
+      if (infer) {
+        WholeProgramInference wpi = atypeFactory.getWholeProgramInference();
+        ExecutableElement methodElt = TreeUtils.elementFromDeclaration(tree);
+        inferPurityAnno(additionalKinds, wpi, methodElt);
+        // The purity of overridden methods is impacted by the purity of this method. If a
+        // superclass method is pure, but an implementation in a subclass is not, WPI ought to treat
+        // **neither** as pure. The purity kind of the superclass method is the LUB of its own
+        // purity and the purity of all the methods that override it. Logically, this rule is the
+        // same as the WPI rule for overrides, but purity isn't a type system and therefore must be
+        // special-cased.
+        Set<? extends ExecutableElement> overriddenMethods =
+            ElementUtils.getOverriddenMethods(methodElt, types);
+        for (ExecutableElement elt : overriddenMethods) {
+          inferPurityAnno(additionalKinds, wpi, elt);
+        }
+        if (overriddenMethods.isEmpty()) {
+          // Handle interfaces. Java overrides by default refer to superclasses,
+          // so only explore the interfaces if no overridden methods were found in superclasses.
+          // Note: ElementUtils#getOverriddenMethods does not return methods in interfaces.
```

Comment:
This is contrary to the method's documentation.  I looked at the implementation, which calls JDK methods, and as of JDK 17 it does seem to include interfaces.  I commented out the containing `if` test and ran tests under JDK 11 and JDK 17, and they passed.

Reply:
You are right. I think I know what happened: there was a bug in my original implementation of the LUB function over purity annotations, and I must have mis-attributed it to this method not including interfaces (because interface methods were apparently not being included, but it just happened that my test for interfaces was the only one that triggered the LUB bug).

I've removed this code.